### PR TITLE
(CDAP-12553) Fix context classloader when calling user method

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java [hunk 1]
```diff
@@ -385,4 +385,4 @@ URI getPySparkScriptLocation() {
   boolean isPySpark() {
     return getPySparkScript() != null || getPySparkScriptLocation() != null;
   }
-}
+}
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java]: accidental change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 2]
```diff
@@ -633,42 +638,81 @@ public void run() throws Exception {
...
+          });
         }
-      });
+      } catch (TransactionConflictException e) {
+        // For Tx conflict, use it as is.
+        throw e;
+      } catch (TransactionFailureException | UncheckedExecutionException e) {
+        // For tx failure (implicit case) or unchecked execution exception (explicit with CDAP wrapper case),
+        // throw the cause for logging.
+        throw e.getCause() == null ? e : e.getCause();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java:672]: seems like all this logic doesn't matter, as it just gets caught as a Throwable below, and we don't do anything with it. Should we at least log the throwable below?

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java:672]: Oh, missing the last argument. Will add it.

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java (hunk 1 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java]: fixed

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-RUT1352-3

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Build  passed https://builds.cask.co/browse/CDAP-RUT1352-8

**Pull request merged**